proxy: Move Kafka proxy to Envoy Go Extensions

jrajahalme opened the pull request:
Move Kafka proxy to Envoy Go Extensions. This has couple of benefits:

- Conceptually simplify Cilium Agent and make it a little bit smaller
  - Removal of some vendored Kafka proxy code now replaced by Envoy/proxylib
- Egress policy support for Kafka proxying
- Performance should be better
  - Envoy threading model is better suited for throughput and low latency, no go-routines are needed
  - Only 8 bytes are parsed from responses, rest is passed on without even passing the data to the parser

The only user-visible change to existing behavior should be the small change in access logging format, and the fact that only CorrelationID is logged for responses.

As an internal implementation detail the generic L7 key-value pair policy format is used to carry Kafka policies from Cilium agent to Envoy. Correspondingly the generic L7 access log format is used to emit the access log messages from the Kafka parser.

```release-note
Kafka proxy is now implemented in Cilium Go extensions for Envoy,
which adds egress policy enforcement support for Kafka L7 policies.
```

jrajahalme committed:
proxylib: Honor L3-dependency also without L7 rules.

This is a proxylib fix to the same bug that was already fixed on
Cilium Envoy filters by
https://github.com/cilium/proxy/commit/c31482c3e49670980c05cafc914320f7949b266f.

Remove the shortcut added earlier for sidecar support to check L3
dependency also when there are no L7 rules. Sidecars should still work
as expected.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

gandro commented on pkg/hubble/parser/seven/parser.go:
```diff
@@ -395,11 +396,16 @@ func decodeLayer7(r *accesslog.LogRecord) *pb.Layer7 {
+	// 	return &pb.Layer7{
+	// 		Type:   flowType,
+	// 		Record: decodeL7(r.Type, r.L7),
+	//	}
```
This looks like a leftover?

jrajahalme replied:
More like wishful thinking... it seems to me that the generic L7 key-value pairs are not translated to the monitor API yet?

gandro commented on pkg/monitor/logrecord.go:
```diff
@@ -167,7 +159,6 @@ type LogRecordNotifyVerbose struct {
 	DstIdentity      uint64                     `json:"dstIdentity"`
 	Verdict          accesslog.FlowVerdict      `json:"verdict"`
 	HTTP             *accesslog.LogRecordHTTP   `json:"http,omitempty"`
-	Kafka            *accesslog.LogRecordKafka  `json:"kafka,omitempty"`
```
This is technically a breaking change in the monitor API. Not opposed to this, but just wanted to point that out explicitly, e.g. legacy versions of Hubble (v0.5) would not be able to recognize Kafka records after this (we don't support Hubble v0.5 on Cilium 1.9, so that's fine).

jrajahalme replied:
Good to know, will leave this for further discussion.

nebril replied:
I think it's fine to go through with that, but would be nice to have a release note. If it's necessary to leave this, can we have information about Kafka in two places (AFAIU right now it will be stored in `L7` field?).

jrajahalme replied:
I'll check how hard it would be to keep using `LogRecordKafka`, will report back!

gandro reviewed (approved):
Hubble-related changes look good to me! One comment with regards to the monitor API.

I did not review the policy or parser changes.

jrajahalme committed:
test: Kafka egress K8s test

Modify exising Kafka K8s test to include an egress Kafka policy while
opening up the corresponding ingress policy to verify that the
enforcement is indeed done by the egress policy.

Correct language in test code and remove unnecessary `fmt.Sprintf()`
calls.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm commented on pkg/hubble/parser/seven/parser.go:
```diff
@@ -513,24 +519,28 @@ func (p *Parser) getSummary(logRecord *accesslog.LogRecord, flow *pb.Flow) strin
+			ApiVersion:    int32(apiVersion),
+			ApiKey:        l7.Fields["APIKey"],
+			CorrelationId: int32(correlationID),
+			Topic:         l7.Fields["Topics"],
```
Should we have constants for these keys? Also, shouldn't this `decodeKafka` function be in a dedicated file, perhaps `pkg/hubble/parser/seven/kafka.go`?

jrajahalme replied:
Dedicated file or not is not an issue for this PR as I did not introduce the function here.

aanm commented on proxylib/kafka/parser.go:
```diff
@@ -0,0 +1,286 @@
+	for _, l7Rule := range allowRules {
+		var kr KafkaRule
+		for k, v := range l7Rule.Rule {
+			switch k {
```
Should we have `switch strings.ToLower(k)` {?

jrajahalme replied:
Not necessarily as these keys are generated by own code in Cilium agent which currently uses all lowercase letters.

aanm commented on vendor/modules.txt:
```diff
@@ -475,7 +475,6 @@ github.com/opencontainers/image-spec/specs-go
 github.com/opencontainers/image-spec/specs-go/v1
```
Given that we are only using the `kafka/proto` we can remove the replace directive for this dependency in `go.mod` and use the one from upstream instead our fork?

jrajahalme replied:
I was not aware of the replace directive, will check!

jrajahalme replied:
@aanm The three commits we have on our fork a head of the upstream ancestor are all on `kafka/proto`. Now that we are not parsing requests at all some of there commits may not be needed any more, but the latest commit fixes an issue in request parsing that seems like we'd want to keep.

aanm commented on proxylib/kafka/parser.go:
```diff
@@ -0,0 +1,286 @@
+				"APIVersion":    strconv.Itoa(int(req.GetVersion())),
+				"APIKey":        apiKeyToString(req.GetAPIKey()),
+				"CorrelationID": strconv.Itoa(int(req.GetCorrelationID())),
+				"Topics":        strings.Join(topics, ","),
```
these could be constants as well

jrajahalme replied:
Will await for some more feedback on the general approach of using the "generic L7" format to internally carry Kafka log messages before doubling down on it :-)

nebril commented on pkg/kafka/policy.go:
```diff
@@ -141,15 +141,14 @@ func matchOffsetFetchReq(req *proto.OffsetFetchReq, rule api.PortRuleKafka) bool
-		fieldRule:    rule,
-	}), "Matching Kafka rule")
+	if flowdebug.Enabled() {
+		log.Debugf("Matching Kafka request %s against rule %v", req.String(), *rule)
```
Why the change to logging method? Seems like structured log was nicer, would it be possible to get back to it?

jrajahalme replied:
Short answer: to avoid confusing log messages.

Long answer: Kafka logs become part of Envoy logs when Kafka is moved to the proxylib. Envoy log messages are not structured and currently we have no code to try parse the structured items from the main log message. This means that if we use structured logs here, they will be embedded in to the main log message and will not be found by any code that will try to parse structured logs. For reference please see https://github.com/cilium/cilium/blob/master/pkg/envoy/envoy.go#L282-L343

nebril commented on pkg/kafka/request.go:
```diff
@@ -217,12 +223,15 @@ func ReadRequest(reader io.Reader) (*RequestMessage, error) {
 	default:
-		log.WithField(fieldRequest, req.String()).Debugf("Unknown Kafka request API key: %d", req.kind)
+		if flowdebug.Enabled() {
+			log.Debugf("Unknown Kafka request API key: %d in %s", req.kind, req.String())
```
Same as comment above, please consider using structured log, (but `if flowdebug.Enabled` line is probably a good call).

Will not comment on more changes to structured logs to not clutter the review.

jrajahalme replied:
See above :-)

nebril reviewed (approved):
Aside from logging nitpicks I don't have any issues. Great work!

aanm reviewed (approved):
Nothing else besides what Maciej already point it out.

aditighag commented on pkg/hubble/parser/seven/parser.go:
```diff
@@ -395,11 +396,16 @@ func decodeLayer7(r *accesslog.LogRecord) *pb.Layer7 {
 			Record: decodeHTTP(r.Type, r.HTTP),
 		}
-	case r.Kafka != nil:
+	case r.L7 != nil && r.L7.Proto == "kafka":
```
minor: Should we extract "kafka" from here and L 511 to a constant?

jrajahalme replied:
The most recent version got rid of the changes in hubble code, so this is not an issue any more :-)

aditighag commented on proxylib/kafka/parser.go:
```diff
@@ -0,0 +1,206 @@
+	req.RequestMessage, err = kafka.ReadRequest(reader)
+	if err != nil {
+		if flowdebug.Enabled() {
+			log.WithError(err).Warning("Unable to parse Kafka request; closing Kafka connection")
```
Does it print metadata (e.g., correlation ID) for the request in the log output?

jrajahalme replied:
Did not, added. For now the errors use the same code as (policy) drops, will need to consider if that may cause confusion.

aditighag commented on proxylib/kafka/parser.go:
```diff
@@ -0,0 +1,206 @@
+	if err != nil {
+		logEntry.GenericL7.Fields["status"] = strconv.Itoa(int(kafka.ErrInvalidMessage))
+		if flowdebug.Enabled() {
+			log.WithError(err).Warning("Unable to create Kafka response")
```
Same comment as L 163.

jrajahalme replied:
Yes, but I may have changed this since you reviewed.

jrajahalme committed:
test: Validate Kafka access logging.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxylib: Add a parser interface for io.Reader parsers.

Kafka request parsing uses io.Reader. Refactor the Reader code away
from the Kafka parser to simplify the parser and to make other
io.Reader parsers easier to integrate.

Kafka replies need minimal parsing, just the first 8 bytes, after
which the whole frame can be passed on. Use the parser frame length to
ask for full frames before parsing requests.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on pkg/policy/api/kafka/doc.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017 Authors of Cilium
+// Copyright 2016-2020 Authors of Cilium
```
nit: 
```suggestion
// Copyright 2017-2020 Authors of Cilium
```

joestringer reviewed (approved):
policy/docs bits LGTM. Awesome that we can reduce code AND add features here (add egress support). Probably even worth release-noting that.

jrajahalme committed:
test: Kafka egress runtime test

Modify one of the existing Kafka tests to use egress enforcement and
visibility on ingress at the same time.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxy: Move Kafka proxy to Envoy Go extensions

Move Kafka proxy from Cilium agent to Envoy Go extensions.

Avoid importing the whole api to proxylib by moving kafka to it's own
api/kafka package. This helps reduce the proxylib binary bloat due to
Kafka from 11.5MB down to 0.3MB.

Move Kafka rule internals from pkg/policy/api/kafka to pkg/kafka. Add
Rule type to pkg/kafka that can also be used from the proxylib kafka
parser as-is.

Simplify pkg/kafka logging to avoid duplicating dates, levels,
etc. when the logs are piped to cilium agent.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
envoy: Update API for Kafka

Add access log message type for Kafka. Split multiple topics to
separate access log messages on the cilium agent side for backwards
compatibility.

Refine the so far unused Kafka rule message.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
